crates/biome_service/src/file_handlers/json.rs

```diff
@@ -179,9 +179,11 @@ fn parse(
     let overrides = settings.map(|s| &s.override_settings);
     let optional_json_file_source = file_source.to_json_file_source();
     let options = JsonParserOptions {
-        allow_comments: parser.map(|p| p.allow_comments).unwrap_or_default()
+        allow_comments: parser.and_then(|p| p.allow_comments).unwrap_or_default()
             || optional_json_file_source.map_or(false, |x| x.allow_comments()),
-        allow_trailing_commas: parser.map(|p| p.allow_trailing_commas).unwrap_or_default()
+        allow_trailing_commas: parser
+            .and_then(|p| p.allow_trailing_commas)
+            .unwrap_or_default()
             || optional_json_file_source.map_or(false, |x| x.allow_trailing_commas()),
     };
```

Comment:
`unwrap_or_default` will cast `None` to `false`, we should do this instead:

```rust
let options = JsonParserOptions {
    allow_comments: parser.and_then(|p| p.allow_comments).map_or_else(
        || optional_json_file_source.map_or(false, |x| x.allow_comments()),
        |value| value,
    ),
    allow_trailing_commas: parser.and_then(|p| p.allow_trailing_commas).map_or_else(
        || optional_json_file_source.map_or(false, |x| x.allow_trailing_commas()),
        |value| value,
    ),
};
```

The logic: we first check the parser options, if they're unset, we fallback to check the file source.